credentials/tls.go

```diff
@@ -53,6 +55,42 @@ func (t TLSInfo) GetSecurityValue() ChannelzSecurityValue {
 	return v
 }
 
+// ParseSpiffeID parses the Spiffe ID from State and fill it into SpiffeID.
+// An error is returned only when we are sure Spiffe ID is used but the format
+// is wrong.
+func (t *TLSInfo) ParseSpiffeID() error {
+	if len(t.State.PeerCertificates) == 0 || len(t.State.PeerCertificates[0].URIs) == 0 {
+		return nil
+	}
+	spiffeIDCnt := 0
+	var spiffeID url.URL
+	for _, uri := range t.State.PeerCertificates[0].URIs {
+		if uri == nil || uri.Scheme != "spiffe" || uri.Opaque != "" || (uri.User != nil && uri.User.Username() != "") {
+			continue
+		}
+		// From this point, we assume the uri is intended for a Spiffe ID.
+		if len(uri.Host)+len(uri.Scheme)+len(uri.RawPath)+4 > 2048 ||
+			len(uri.Host)+len(uri.Scheme)+len(uri.Path)+4 > 2048 {
+			return fmt.Errorf("invalid SPIFFE ID: total ID length larger than 2048 bytes")
+		}
+		if len(uri.Host) == 0 || len(uri.RawPath) == 0 || len(uri.Path) == 0 {
+			return fmt.Errorf("invalid SPIFFE ID: domain or workload ID is empty")
+		}
+		if len(uri.Host) > 255 {
+			return fmt.Errorf("invalid SPIFFE ID: domain length larger than 255 characters")
+		}
+		// We use a default deep copy since we know the User field of a SPIFFE ID is empty.
+		spiffeID = *uri
+		spiffeIDCnt++
+	}
+	// A standard SPIFFE ID should be unique. If there are more, we don't raise
+	// any errors but simply not plumbing any of them.
+	if spiffeIDCnt == 1 {
```

Comment:
Having two SPIFFE IDs is wrong. We should return error or at least log error. Otherwise, it hard to know what goes wrong if there are two SPIFFE IDs in a cert.

PS: We should also log error in C core implementation.

Reply:
Done. For go version <= 1.9, parsing of URI is not supported by the underlying lib so I just don't not support it, using go's build constraints. 
Please take a look and let me know if you have any questions. Thanks!